apollo-router/src/plugins/authentication.rs

```diff
@@ -189,6 +204,136 @@ async fn get_jwks(url: Url) -> Option<JwkSet> {
+                // If a key doesn't have an algorithm, then we match the "alg" specified in the
+                // search criteria against all of the algorithms that we support.  If the
+                // key.algorithm parameters match the type of parameters for the "family" of the
+                // criteria "alg", then we update the key to use the value of "alg" provided in
+                // the search criteria.
+                // If not, then this is not a usable key for this JWT
+                // Note: Matching algorithm parameters may seem unusual, but the appropriate
+                // algorithm details are not structured for easy consumption in jsonwebtoken and
+                // this is the simplest way to determine algorithm family.
+                None => match criteria.alg {
+                    Algorithm::HS256 | Algorithm::HS384 | Algorithm::HS512 => {
+                        if let AlgorithmParameters::OctetKey(_) = key.algorithm {
+                            key.common.algorithm = Some(criteria.alg);
+                        } else {
+                            continue;
+                        }
+                    }
+                    Algorithm::RS256
+                    | Algorithm::RS384
+                    | Algorithm::RS512
+                    | Algorithm::PS256
+                    | Algorithm::PS384
+                    | Algorithm::PS512 => {
+                        if let AlgorithmParameters::RSA(_) = key.algorithm {
+                            key.common.algorithm = Some(criteria.alg);
+                        } else {
+                            continue;
+                        }
+                    }
+                    Algorithm::ES256 | Algorithm::ES384 | Algorithm::EdDSA => {
+                        if let AlgorithmParameters::EllipticCurve(_) = key.algorithm {
+                            key.common.algorithm = Some(criteria.alg);
+                        } else {
+                            continue;
+                        }
+                    }
+                },
+            };
+
+            // If we get here we have a key that:
+            //  - may be used for signature verification
+            //  - has a matching algorithm, or if JWT has no algorithm, a matching key type
+            // It may have a matching kid if the JWT has a kid and it matches the key kid
+            //
+            // Multiple keys may meet the matching criteria, but they have a score. They get 1
+            // point for having an explicitly matching algorithm and 1 point for an explicitly
+            // matching kid. We will sort our candidates and pick the key with the highest score.
+
+            // If we find a key with a HIGHEST_SCORE, let's stop looking.
+            if key_score == HIGHEST_SCORE {
+                return Ok(Some(key));
+            }
+
+            candidates.push((key_score, key));
+        }
+    }
+    tracing::debug!("jwk candidates: {:?}", candidates);
```

Comment:
it's not a good idea to log the entire key structure, as it can contain a symmetric key. `kid`, `kty` and `alg` should be enough?

Reply:
I'm now logging : score, optional kid, algorithm

done: 53b6295c1934a75d51e24fcb8f198576af8f7fb4